### DISCUSSION TITLE
Frigate Event Statistics and Analytics

### DISCUSSION THREAD

**Discussion by moandcompany** [blakeblackshear/frigate#3676] [Ideas]
Hi,

I was wondering if there has been any past discussion or interest in having an event analytics layer for Frigate.

An analytics layer could be used to generate event statistics to help Frigate users understand the frequency of events, by camera, zone, object-type, time of day, etc. An analytics layer can also be used to help focus and expedite work needed for Frigate detector model development by identifying events and captured image/video data associated with false-positives and misclassifications.

My initial thoughts and approach to this take advantage of the existing Frigate database structure implemented using SQLite, where the database can be mounted in read-only mode and exposed via [Datasette](https://datasette.io). Datasette includes functionality to incorporate canned SQL queries/reports, as well as exposing Frigate data via REST API.

With Frigate event data organized in an easily queryable manner, this would also provide a path for implementing event dashboards summarizing events over time.

This could also be used as part of a development and test harness to expose and compare performance between Frigate configuration settings and detector models.

**Comment by moandcompany** [comment]
Adding a note for anyone else with similar interest

A general event summary is available using Frigate's existing API, [documented here](https://docs.frigate.video/integrations/api#get-apieventssummary), implemented through route: [/api/events/summary](https://github.com/blakeblackshear/frigate/blob/6e2e297aeb451598223a3beb6735eb5fc27b37b2/frigate/http.py#L75)

The summary provides a daily count of event labels, by camera, label class, and zone

**Comment by moandcompany** [comment]
![timeline](https://user-images.githubusercontent.com/1381775/188085602-08b27d75-901e-4c5d-9d6a-917ad4fba0be.png)

Something I'm interested in seeing with Frigate is a timeline view of an individual day, and multiple days, to better understand where events occurred, if there are activity patterns, spot activity when it is most frequent, and when it occurs but is an anomaly (e.g. late-night hours).

**Comment by moandcompany** [comment]
![labels](https://user-images.githubusercontent.com/1381775/188214520-c4bf9fff-55a8-4aa6-9b47-47e2eff6164e.png)

Another useful view would be for labels, to analyze event detection frequency, and some their attributes (e.g. confidence vs detection area)

**Comment by grzywek** [comment]
Wow those examples look really cool.

**Comment by manustars** [comment]
this is good, i integrate frigate on Homeassistant, i would see the event statics in a card (only number for day) but the integration not see any sensor for "events statics"

**Comment by moandcompany** [comment]
Bumping this thread in light of the Frigate 0.14.x UI design changes.

I've been working on my own UI for reviewing Frigate events as one of my use cases diverges from the primary case of recording very-recent Detections and reviewing Notifications/Alerts.

The 0.14.x UI approach understandably prioritizes simplification of reviewing general "alerts" as the Frigate approach in 0.13.x and earlier presented detection information separately for each object/label class even when multiple types of objects were part of the same actual "event."

Some Frigate users may have cameras covering relatively high-activity areas (e.g. people, vehicles, etc) and in those cases they may be:

1. Figuring out how to configure the system to minimize detection/event recordings to only particular areas (e.g. through use of zones) -> in order to minimize the number of events to "review" or to minimize the number of notifications through an associated integration (e.g. via Home Assistant).

2. Configuring the system to record most of, if not everything, detected -> in order to enable an indexed review of the past as part of a broad review or analysis of historical activity in a particular area.

Both of these have tradeoffs and require thoughtful consideration of how Frigate itself would be integrated into a larger workflow.

![timeline](https://github.com/blakeblackshear/frigate/assets/1381775/498803e8-cd56-4a35-9011-f77b44034324)

The example chart above shows an overview of 30-days and how detection events were distributed within each day (by hour).

I believe it's important to be able to have alternate views of Frigate event/detection records beyond the immediate and granular timelines of the last few hours, or the last day, and to have a broader-scale view of time allowing easy observation and discovery of events that may be temporally anomalous (e.g. detections occurring during late-night hours, difference in traffic patterns on particular days of the week).

Some example questions that could be answered with these views are:
- What periods of a day are most busy? Versus when is an event out of norms?
- How do weekdays versus weekends differ in terms of traffic patterns?
- What are the most common modes of transport being used (e.g. on foot, bike/motorcycle, car)?

**Comment by ofcedwards** [comment]
I like the data analytics, but I think this really diverges from what most people want to get out of an NVR. Especially considering that most people I've run into with a proprietary NVR hardly even know how to play back video, I doubt the average user cares about visualizing data. I can agree that the average frigate user is probably a little more into tech, but I think the point still stands.

It's also worth noting that a lot of data is available through the api. I use api calls to graph data I care about in home assistant. This type of data analysis might be better included as its own project in the form a home assistant integration that can be used those interested in it.

**Reply by VideoFX** [reply]
Disagree. The stats are a huge part of NVR, especially one with object recognition, because you can understand the events better. For example, you can visualize clusters of events, such as cars or people, to understand time of day effects. Many more things like this. Normally you would just "eyeball" it, but with this data, you can truly understand your events over time, and see things you would have never otherwise seen. This enlightenment is something frigate has brought us already, so it would be a logical roadmap I think.

That said, it doesn't have to be a frigate feature, it could be an addon or  script or something, but it would be a powerhouse as a feature. Just saying it fits.
